### ISSUE TITLE
Bad Image MaxTo.Hook.x86.dll

### ISSUE THREAD

**Issue by khongi** [digitalcreations/MaxTo#415]
**Describe the bug**
Whenever I switch to Spotify, I get an error popup with the following message:

`Spotify Premium: Spotify.exe - Bad Image : C:\Users\Khongor\AppData\Local\MaxTo\app-2.0.1\MaxTo.Hook.x86.dll is either not designed to run on Windows or it contains an error. Try installing the program again using the original installation media or contact your system administrator or the software vendor for support. Error status 0xc000012f. `

MaxTo keeps working after the error message, but when the Spotify is in focus the error popup reappears immediatly.

**To Reproduce**
Whenever Spotify is focused, Just installed the 2019 May update, not sure if relevant.

**Screenshots**
![image](https://user-images.githubusercontent.com/11983684/58277211-7a7d2180-7d99-11e9-9051-b71b59d66f05.png)

**System information:**
 - Windows version: 10 2019 May Update
 - MaxTo version 2.0.1

**Additional context**
I started MaxTo, reproduces the error and stopped it. Attached the relevant log part.
[maxto_log_start_error_stop.log](https://github.com/digitalcreations/MaxTo/files/3213647/maxto_log_start_error_stop.log)

**Comment by vegardlarsen** [comment]
So I tested this with Spotify version 1.1.4.197.g92d52c4f and 1.1.7.13766.gf9dc3904 on Windows 10 build 1809. That appears to work just fine.

Which version of Spotify do you have installed currently?

**Comment by martillodecampo** [comment]
I also received similar error dialogs referencing AppleMobileDeviceProcess.exe and SkypeBridge.exe immediately after installing the Feature update to Windows 10, version 1903 earlier this weekend. Windows Update also applied KB4497932, KB4495620, KB2994397 and KB2992080 at the same time. The current system version is now 1903.18362.116. I had to uninstall MaxTo in order to be able to use my PC.

![image](https://user-images.githubusercontent.com/22117756/58340337-9f18de00-7e19-11e9-94ef-dda0519d3135.png)

![image](https://user-images.githubusercontent.com/22117756/58340364-b2c44480-7e19-11e9-84de-f9879112fff6.png)

**Comment by khongi** [comment]
My Spotify version is 1.1.7.13766.gf9dc3904 (Windows Store version)

**Comment by khongi** [comment]
I reinstalled Spotify, seems like it solved the issue

**Comment by poita66** [comment]
Just updated to Windows 10 1903 and now have this issue with Spotify and Dell Update.

**Comment by JCbarreau** [comment]
I have the same problem with Amazon Music App. 
Each time I want to clic on it, it makes this error appears around 20 times!

I must say that I am part of Program WIndows Insider, and this error is now appearing for around a month.

**Comment by fsch** [comment]
startet to show up after upgrade to 1903 on my side. I think it is more of a WIN10 1903 issue than a spotify issue.

**Comment by morgdig** [comment]
This error started after upgrade to 1903. Do not have spotify.

**Comment by vegardlarsen** [comment]
It would appear this is more a Windows 1903 issue than a Spotify issue, indeed. It is hard to debug, as after updating my own machine to 1903 I haven't seen this happen even once. I suspect attaching a debugger to a program doing this would reveal what the problem is quite quickly.

**Comment by JCbarreau** [comment]
In fact this is a Windows issue. It appears also with skypeBridge. For me, you should verify with app installed (not programs installed, but app from Microsoft store). I am quite sure this comes from there

**Comment by tynril** [comment]
I installed Windows 10 1903 today, and this error started happening right after first boot. I had it happen with a few programs, Spotify, SkypeBridge, and GameBar.

That Windows update is getting released to the general public at the moment, so it is likely more users will start experiencing this.

**Comment by vegardlarsen** [comment]
I am aware that 1903 is being rolled out generally. There does not appear to be anything we can do to fix the problem:

1. It is not reproducible (it only happens on some machines after an upgrade, not on others)
2. The problem goes away by re-installing the crashing application, not MaxTo

For us to be able to fix an issue in MaxTo, it has to at a minimum be reproducible on one of our machines. After upgrading a few of them to 1903, we still haven't seen this happen.

**Comment by vegardlarsen** [comment]
Just for context here, you get "bad image" errors for a multiude of reasons. Usually they are caused by either corrupt files, or if you are trying to load a 64-bit DLL into a 32-bit process or the other way around.

The error status referenced (`0xC000012F`) is `STATUS_INVALID_IMAGE_NOT_MZ`, and the description is "The specified image file did not have the correct format: it did not have an initial MZ.". [Documentation](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-erref/596a1078-e883-4972-9bbc-49e60bebca55)

`MZ` in this context means the two ASCII bytes `4D 5A`, which identifies a DOS executable. My interpretation here is that the DLL is attempted to be loaded, and it can't find out which kind it is, and in a last ditch effort tries to load it as a DOS executable. My interpretation here can absolutely be wrong; something completely different may be happening here.

Still no progress on actually fixing this.

**Comment by baabo** [comment]
I get this error with Telegram installed from Microsoft Store. I noticed that the error only pops up when MaxTo is in active mode.

**Comment by vegardlarsen** [comment]
@baabo That makes sense. MaxTo only loads its DLL into other programs when MaxTo itself is active.

**Comment by poita66** [comment]
For those coming here for a quick solution: 'Reset' the broken app in the 'Settings' app.
#### BEWARE: this will effectively reinstall the app and you'll lose your data for that app and need to login to the app again
To do this:
1. go to Apps & Features (right-click the start menu button, it's at the top)
2. find the app in question (e.g. Spotify)
3. press the 'Reset' button

**Comment by JotaEleEme** [comment]
Same issue running windows 1903... No spotify or similar application running. After reinstall, MaxTo works fine for a day, and then starts to show the "bad image" error.
![Capture](https://user-images.githubusercontent.com/52716352/60912418-2fea3280-a27d-11e9-8f91-f7ed6a90333b.PNG)

**Comment by vegardlarsen** [comment]
@JotaEleEme in your case this seems to be caused by LocalBridge.exe. I don't know what kind of application that is, but reinstalling or resetting it should work.

**Comment by Ryknyk** [comment]
I have this issue too, after upgrading windows to 1903, error dialog appears for a variety of apps multiple times, I'm not going to be re-installing everything else. Clearly this bug relates to Max.to. Very frustrating, I have to kill Max.to on each boot, haven't been able to leave it active for a few weeks now.

**Comment by vegardlarsen** [comment]
@Ryknyk Absolutely this bug "relates" to MaxTo, but I am not sure it is MaxTo that does anything wrong in this case. Considering that the application starts working after a reset or reinstall, and that this issue is not reproducible on new installations of Windows 10 1903 or on all upgrades to Windows 10 1903, it is very hard to find out what is actually happening here. MaxTo relies on Windows to load its DLL into other processes, and here it appears that Windows is loading the wrong "bitness" of DLL.

**Comment by altano** [comment]
> and here it appears that Windows is loading the wrong "bitness" of DLL.

This isn't a bitness mismatch:
![image](https://user-images.githubusercontent.com/1009/62735459-148b5800-b9e0-11e9-935d-47727315eb77.png)

**Comment by altano** [comment]
@vegardlarsen I dug into this a bit more and I'm noticing something interesting:

The MaxTo.Hook.x64.dll that is trying to be loaded is NOT the one that sits in the MaxTo folder. For example I have a Windows Store app called QuickLook that is receiving this error. The DLL it is trying to load is cached in the QuickLook folder:

```
C:\Users\alan\AppData\Local\Packages\21090PaddyXu.QuickLook_egxr34yet59cg\LocalCache\Local\MaxTo\app-2.0.1\MaxTo.Hook.x64.dll

Name: MaxTo.Hook.x64.dll
Size: 127984 bytes (124 KiB)
SHA256: 57582261DCC2C41C8304169FF21EE59C90EEEE377C715661AA8014BC0D92307E
```
Meanwhile the MaxTo folder has this DLL which is NOT the same (note the different SHA256 CRC):
```
C:\Users\alan\AppData\Local\MaxTo\app-2.0.1\MaxTo.Hook.x64.dll

Name: MaxTo.Hook.x64.dll
Size: 127984 bytes (124 KiB)
SHA256: BA38651E2C8F31619E707BA59AF15CFD5A93E51C8AD6874807700DDDEB491CA3
```

I took one of these apps that was hitting this problem and did as someone in this thread suggested and ran "Reset" in the Windows Settings for the app. When you run Reset it is deleting the app's `%LocalAppData%\Packages\<APP PACKAGE>\LocalCache\Local` folder (including the MaxTo subfolder). Then when you re-run the app it recreates the LocalCache\Local folder and it gets populated with some other things (e.g. a Microsoft subfolder) but the MaxTo directory is NOT recreated.

If you don't want to manually Reset all the apps hitting this error you can get the same outcome by:
1. Navigating to the `%LocalAppData%\Packages` folder
1. Searching for `kind:folder name:maxto`
1. Deleting every result

This will fix the problem with the error dialog popping up, but note that I verified in Process Monitor that under the hood these apps are still looking for the MaxTo.Hook.x64 DLL: it just can't find it, and since it can't find it we never get a visible error dialog. It just produces an error under the covers.

I've never developed Windows Store apps so I don't know what these cached DLLs are, how they got there, how they became corrupt, or the right way to delete them. Perhaps someone else can pickup the trail from here?

**Comment by altano** [comment]
And FWIW replacing these cached MaxTo DLLs with the %LocalAppData%\MaxTo\app-2.0.1\MaxTo.Hook.x64.dll one just results in the apps crashing on startup. So the only workaround I can see is to delete the cached ones as I suggested above.

**Comment by vegardlarsen** [comment]
@altano This is extremely interesting. On my phone at the moment, but could you send me a copy of this modified file at the support address?

**Comment by altano** [comment]
@vegardlarsen LOL I zipped it up to send it to you and the Zip compressed it from 125KB to 1KB. I peeked inside and the DLL is just 125KB of 0s: https://files.terriblefish.com/MaxTo.Hook.x64.dll.zip

What on earth?

**Comment by vegardlarsen** [comment]
I'll probably open a support issue with Microsoft on this. This sounds extremely strange.

**Comment by altano** [comment]
One more update: I checked every way I could find to see what kind of links (e.g. hard link, reparse point) these folders are and I can't find anything. Maybe they aren't links and I just accidentally did a Select-All in the Explorer search results before doing the folder rename. When you do that and rename one item it renames all of them.

**Comment by pvh** [comment]
Just a quick note to say that this issue is affecting me as well and that as far as I can tell, all the applications mentioned appear to have tray apps associated with them (Skype, Spotify, GameBar, Telegram.)

**Comment by dukeduck1984** [comment]
same here, after upgraded to win10 1903, that error popped up almost evey time I launched an application

**Comment by vegardlarsen** [comment]
Could anyone who is currently experiencing this error, try to locate the file on their system, and query it with `fsutil sparse queryFlag <path>`? @altano

This information is not included when the file is sent over the Internet.

**Comment by altano** [comment]
```
PS C:\> fsutil sparse queryFlag C:\Users\alan\AppData\Local\Packages\AppleInc.iTunes_nzyj5cx40ttqa\LocalCache\Local\MaxTo-old\app-2.0.1\MaxTo.Hook.x64.dll

This file is set as sparse
```

**Comment by vegardlarsen** [comment]
Thank you @altano. That was my suspicion. Uninitialized sparse files are usually zero-filled. I am waiting on feedback from Microsoft on this issue.

**Comment by vegardlarsen** [comment]
I have been able to reproduce this in a virtual machine, and I am looking into it with the help of Microsoft. I will update when I know more.

**Comment by deegardner8** [comment]
I get the error window with a windows store app called "Widgets HD"

**Comment by vegardlarsen** [comment]
@deegardner8 The same procedure as for the other apps should help you clear that out. It should be a one-time thing.

Progress at the moment is that we are being "bounced around" a bit by Microsoft, but I will hopefully have an update on this soon.

**Comment by deegardner8** [comment]
@vegardlarsen Thanks for the suggestion.  It has cleared the error deleting the files out of the cache directories.

**Comment by vegardlarsen** [comment]
Just an update from my work with Microsoft to figure this out. We've found that if a DLL located inside `%LocalAppData%` calls `SetWindowsHookEx` to set up a global CBT hook (which is the mechanism MaxTo uses to intercept e.g. maximize events), specific (maybe all) Windows Store-installed apps will mirror that DLL under its own folder structure in `%LocalAppData%\Packages\<package id>\LocalCache`. The "mirrored" DLL is created as a sparse file with the same file size as the real DLL; but it has no content (if you try to read it, there is only zeroes).

This does not crash the installed Store apps on Windows 10 version 1809, but it does crash them when MaxTo is running on version 1903.

We are working to figure out more details, but we have a minimal reproduction of this issue at this time.

**Comment by vegardlarsen** [comment]
Another update; Microsoft has confirmed that this is a bug in Windows, and have put engineering resources on this. I am still waiting on the details here, and their plans for fixing this issue. I have a meeting scheduled with them tomorrow.

After that, we will probably send out an informational e-mail to all of our customers in case this is something a lot more people are experiencing.

**Comment by vegardlarsen** [comment]
All of our customers should have received the following e-mail earlier today. I am pasting it here for everyone's reference:

> Dear MaxTo customer,
> 
> we are writing to inform you about an issue you may come across when upgrading to Windows 10 version 1903. If you have seen the following error dialog after upgrading to Windows 10, this e-mail is the answer to your problems.
>
> ![image](https://user-images.githubusercontent.com/11983684/58277211-7a7d2180-7d99-11e9-9051-b71b59d66f05.png)
> 
> The issue can cause any application installed through Microsoft Store to crash when started. MaxTo is referenced in the error message, but the error is actually caused by Windows creating a file with MaxTo's name.
> 
> # How to fix it
> 
> In the screenshot above, you can see that Spotify has crashed because of a DLL file named `MaxTo.Hook.x86.dll`. The name of the application is only shown in the title bar of the error.
> 
> ## Resetting application
> 
> The easiest method for solving this issue is to reset the affected application. We are using Spotify as the example here, but these instructions will work for any other application that crashes in this manner.
> 
> Resetting an application in this manner means that the application will lose all its settings. In the case of Spotify, that means you will have to sign in again. In the case of other applications, it may mean data loss. Be aware of this before following our steps.
> 
> 1. Right click the Start menu (or hit Win+X on your keyboard), and open "Apps and features".
> 2. In the window that opens, in "Search this list", type "Spotify".
> 3. In the now very short list, click "Spotify", and then click the "Advanced options" link.
> 4. Scroll down to "Reset", and click the "Reset" button.
>
> After doing this, you will no longer receive the error message for the Spotify application. Repeat for any other application that may be crashing in this manner.
> 
> ## Deleting files manually
> 
> The manual way of fixing this issue requires some more advanced computer knowledge, but can not lead to data loss in the affected applications if done correctly.
> 
> 1. Hit Win+R on your keyboard, and type `%LocalAppData%\Packages`, then click OK.
> 2. The folder you have opened contains all applications installed from the Microsoft Store. Find the folder that corresponds to the affected application. In our case the folder is named `SpotifyAB.SpotifyMusic_zpdnekdrzrea0`. Open that folder.
> 3. Inside this folder, navigate to `LocalCache\Local\MaxTo\app-2.0.1`
> 4. You should find one file, named either `MaxTo.Hook.x86.dll` or `MaxTo.Hook.x64.dll`. Delete that file.
>
> This should prevent this issue from showing up again for the affected application.
> 
> # Who will experience this?
> 
> This error occurs in a very specific set of circumstances:
> 
> 1. You have Windows 10 version 1809 installed on your machine.
> 2. You have installed applications from the Microsoft Store. Known affected applications are (there are probably more):
>    - Spotify
>    - QuickLook
>    - Skype
>    - iTunes (`AppleMobileDeviceProcess`)
>    - Amazon Music App
>    - Dell Update
> 3. You used MaxTo at the same time as an affected application.
> 4. You upgrade to Windows 10 version 1903.
> 5. You experience the crash after the update.
>
> # What causes this issue?
> 
> This error is caused by a change Microsoft made with Windows, starting in version 1903. This section is going to go into a lot of details, as I know a lot of our customers are programmers, and I thought you may find this write-up interesting.
> 
> TL;DR: Microsoft introduced a bug that removes a flag on a placeholder file they created for Microsoft Store apps.
> 
> Before we start on this story, some backstory on how MaxTo works is necessary. To enable MaxTo to listen in on events in your system, we set what is known as a "global CBT hook" (CBT means "computer based training"). To set this hook, MaxTo needs to provide a unmanaged DLL compiled for each platform. These DLLs are called `MaxTo.Hook.x64.dll` and `MaxTo.Hook.x86.dll` respectively.
> 
> When we install the hook, these DLLs are loaded into the applications running on your machine. The DLLs then forward any events they deem to be important enough to the main MaxTo process on your machine.
> 
> When we initially received this error report from our customers, we couldn't reproduce it on Windows 10 version 1903 at all. We tried with fresh installs in virtual machines, I upgraded my computer to 1903, and the error never occurred on our machines. In hindsight this was because I personally didn't use Microsoft Store apps on my machine.
> 
> It was a hint from a customer who experienced the problem, that the path to the hook DLL was not really what it said in the error message, that got us on the right path. He found that the DLL file that was attempted to be loaded was actually inside the Microsoft Store packages folder (a location that MaxTo doesn't use).
> 
> We contacted Microsoft, and after quite a bit of back-and-forth, we managed to produce a minimal reproduction of the issue, basically a simple application that sets a CBT hook in the same manner MaxTo does. We provided this to Microsoft, they put a lot of engineering effort into helping us solve this problem.
> 
> It turns out that Microsoft Store apps run in a sort of "container", meaning they are partially isolated from the rest of the system. This means that they need to have any DLLs they use inside their own folder, so they can access them without breaking out of their sandbox. When you inject a DLL with a CBT hook, the container creates a special [NTFS reparse point](https://en.wikipedia.org/wiki/NTFS_reparse_point) with the name of the DLL you are trying to inject.
> 
> In Windows 10 version 1903, the startup behavior for Microsoft Store containers changed, and due to a bug in Windows, the reparse point flag was removed when starting the application on the updated Windows version.
> 
> Left behind was a DLL file that was empty; it had the correct file size, but if you examined it with an editor you would see that the entire file was zero-filled. Indeed, when our customer zipped the file before sending it to us, the Zip file compressed it from 125kB to 1kB.
> 
> The Microsoft Store container system would try to load this DLL file as if it was a part of the application that you were starting, but would crash when it couldn't make sense of the file.
> 
> Microsoft has now filed an internal issue for this bug, and it will be fixed on their end. We do not know the details of when a fix for this will be available.
> 
> We will continue to update the Github issue when we get any updates from Microsoft on when this fix will be deployed.

Thank you to all of you for helping us track down this issue. Special thanks to @altano, without him we probably wouldn't even have thought in that direction.

I will keep updating this issue if I get more information from Microsoft, but for now I consider this issue solved on our end. Therefore I am closing this issue, but feel free to comment if you have any questions regarding this issue.

**Comment by Xairoo** [comment]
GameBarFT.exe - a new one and I can't find this, just deleted all Xbox folders, that worked.

This problems is a really big thing. **Steals too much (life)time**.

**Comment by vegardlarsen** [comment]
@Xairoo I know. Microsoft has not yet indicated to me when this will be fixed in a Windows update or hotfix, but I know it is on their list of things that will be fixed.

**Comment by Xairoo** [comment]
@vegardlarsen Oh yeah I forgot to ask... How frequently do you ask Microsoft about updates? Weekly? That would be a good start. Can't read anything about making some pressure on Microsoft. If there isn't anything you guys could do, just keep asking and let them know that people are dying because you can't sell new licenses or people run out of lifetime ;-)

**Comment by vegardlarsen** [comment]
@Xairoo As long as we have the above workaround; this probably isn't a critical error on Microsoft's end. We do have a support contract, and we will get updates once something is decided.

And I tend to agree with their assessment, even though this issue affects us very hard, it is a bug that only affects a very limited range of programs (from Microsoft's point of view). While it is definitely something I think they should fix (and they are fixing it), putting pressure on someone in this case would be unhelpful. I am confident they are on it and it will be fixed.

They have been very helpful in reproducing and diagnosing this issue, and I have no doubt they will be just as diligent in fixing it.

**Comment by jneuhaus20** [comment]
Clearly this happened again with 1909, and I wanted to clarify some things for others that weren't immediately obvious to me, especially when the issue presents itself with LocalBridge.exe.

1. You may need `procexp` (from the SysInternals Suite) to figure out what exactly needs fixing.
1. LocalBridge is a Windows app thing that a different app will launch in the course of doing its business.
1. Your error message may give the path to the _correct_ dll, not the problematic one (see image.)
1. Using `procexp` you can find LocalBridge.exe to determine what app is using it, and view the file handles to see the location of this sparse file (see image)

![image](https://user-images.githubusercontent.com/38505603/70646308-3f92ec00-1c0c-11ea-9f6a-e9dfa7635bef.png)

**Comment by vegardlarsen** [comment]
@jneuhaus20 Did this happen to you when updating from 1903 to 1909, or did you update from 1809 to 1909?

I have only seen `LocalBridge.exe` as part of the Microsoft Office Hub-package, and resetting that should work.

**Comment by giorgoc** [comment]
How can we reset `LocalBridge.exe` ? I tried to locate it using the steps above (Win+X, Apps & Features, etc) but apparently I cannot find it onto that list.

**Comment by vegardlarsen** [comment]
@giorgoc You can uninstall that package by running the following PowerShell command (as administrator):

```
get-appxpackage | ? {$_.packagefullname -like '*MicrosoftOfficeHub*'} | remove-appxpackage
```

I have yet to see anyone mention the name of this package, and it does not appear on any of my machines, so I don't know exactly what it is called in _Add or remove programs_.

**Comment by giorgoc** [comment]
@vegardlarsen after I do this , how can I reinstall it if we don't know the exact name? Same command above, but add-appxpackage ?

**Comment by vegardlarsen** [comment]
@giorgoc I don't know. If you cut off the `| remove-appxpackage` part of the command above, you'll get all the information about the package. I would have done this myself, except my machines apparently doesn't have this package. If you do this, please copy your output here so I can examine it.

You can re-install all default Windows packages by running the following command. Not sure if this package is one of those:

```powershell
Get-AppxPackage -AllUsers| Foreach {Add-AppxPackage -DisableDevelopmentMode -Register "$($_.InstallLocation)\AppXManifest.xml"}
```

**Comment by giorgoc** [comment]
@vegardlarsen I tried the above, waited for a couple of days to see if something went missing when removed `LocalBridge.exe` through the suggested method above. So far so good i would say.

So if others have the same, here's the two Powershell commands to use (as admin, always!)

`get-appxpackage | ? {$_.packagefullname -like '*MicrosoftOfficeHub*'} | remove-appxpackage`

and then, when done, this 

`Get-AppxPackage -AllUsers| Foreach {Add-AppxPackage -DisableDevelopmentMode -Register "$($_.InstallLocation)\AppXManifest.xml"}`

**Comment by tolgabalci** [comment]
MaxTo is now unusable for me on Windows 10.  The workaround listed is worthless.  Every time I reset or uninstall an app another one crashes with the same Bad Image error.  Spotify, then Skype (LocalBridge.exe), then AdobeNotificationHelper.exe, and the list keeps going.

**Comment by jneuhaus20** [comment]
@vegardlarsen 1903 to 1909. Deleting the offending DLL resolved it.  I have Spotify installed, but only saw it with LocalBridge.

@tolgabalci Worthless means no worth was provided to anyone, which is clearly not true as it works for some, so false by definition. I get that it's frustrating, but dial back the hyperbole.  Try just deleting the DLL clones instead, you might be in a state where reinstalling is just recreating the error state.

**Comment by vegardlarsen** [comment]
@tolgabalci While I understand that this is frustrating, I am confident that this problem will be solved for you as well once you "get through" the affected applications on your machine. You may just have been particularly unlucky in which applications you have installed.

**Comment by tauCete** [comment]
@vegardlarsen , Thank you very much.  Running repair on the 'Xbox Game Bar' app seems to have fixed the problem for me.  just fyi for anyone else who runs into the problem, my game bar version is 3.31.2003.0 and windows version is 1909 (18363.535) and the issue started happening after a windows update.

**Comment by vegardlarsen** [comment]
Short update from Microsoft: they are taking this fix to "shiproom" this week. I'll get sometime later this week regarding what form the fix will take, and hopefully when it will be made available.
